### DISCUSSION TITLE
Secure local instance through custom Ngrok domain

### DISCUSSION THREAD

**Discussion by ammonhaggerty** [laravel/valet#1476] [Q&A]
I'm having trouble with this and hoping it's possible. Steps I'm taking:
1. Create link (`valet link dev`)
2. Secure link (`valet secure`)
3. Share with Ngrok custom domain (`valet share --domain=dev.mydomain.com`)

When I do this, I see in Session Status:
`Forwarding    https://dev.mydomain.com -> http://dev.test:60`
I'd like to be seeing `-> https://dev.test:80`
I've tried `valet share secure --domain=dev.mydomain.com 80` and other various things I can think of and no love.
Ngrok deployed a TLS cert to dev.mydomain.com, so there's that.
Any other ideas?

**Comment by drbyte** [comment]
> `Forwarding    https://dev.mydomain.com -> http://dev.test:60`

Yes, that's what you should see.

> I'd like to be seeing `-> https://dev.test:80`

Why do you think you need `:80` ? 

It's specifically set to share on port 60 so that there isn't a clash with something already on port 80. And your Valet site's nginx config is configured to listen on port 60 for shared instances.

Also, it'd be extremely rare to see `https` with port `80`. Some webserver engines will actually halt with errors when configured that way.


The way it works is: 
- browser hits `https://dev.mydomain.com`
- ngrok responds to the domain and processes the https certificate validating the https request
- ngrok then proxies the call by telling the browser that it's actually served on `http://dev.test:60`
- your browser looks up `dev.test` locally and your dnsmasq responds with your local IP
- your browser sees the port `:60` and therefore makes the local call to your localhost IP on port 60
- Valet's nginx config listening for `dev.test` on port `60` responds and serves your site

> "I'm having trouble with this".

What exactly is the trouble?

What's it "not doing" that you're expecting? 

What is it "doing" that you're not expecting?  Some sort of error?

**Reply by ammonhaggerty** [reply]
@drbyte - Thank you for the clear explanation! I was just parroting something I read re port 80 routing to https. But was you shared makes sense. What I'm seeing are "mixed content" errors — what appears to be a lack of enforcement for serving through https. 

BTW, when I use the ngrok domains (e.g., myname.ngrok.io) in the same config, everything works. When I add a subdomain of my own domain, that's where I'm seeing issues.

Is there a way to enforce https in this Valet configuration?

**Reply by drbyte** [reply]
When your browser says "mixed content error", does it tell you what the URL was for the non-secure content? Is that URL the same as your secure content but just without `https://` ?  Or does it use a different URL, a different domain?

**Reply by ammonhaggerty** [reply]
Sorry for the slow reply. I had to move on and figure another way to do it. I'm guessing there's some issue with how Valet connects to a personal domain (https://my.domain.com) in ngrok since everything works fine when it's connected in the same way to a custom domain (https://custom.ngrok.io). In both cases, the domains have been issued SSL certs, but personal domain just throws tons of "mixed content errors" — mostly from 3rd party libraries hitting CORS errors (I think). I'd love to resolve it, but perhaps I'm the only person trying to use a custom/personal domain with Valet and ngrok.
